mitodl/mitxpro
# fix: prevent HTML/URLs in the Full Name field

### What are the relevant tickets?
[#4211 ](https://github.com/mitodl/hq/issues/4211)


### Description (What does it do?)
This PR adds validation for the full name. It also introduces an API to validate registration on Open edX to ensure a valid full name is provided. This change affects both the registration details page and the edit profile page.


### Screenshots (if appropriate):
Validation error on registration details page:
<img width="1280" alt="Registration details page 1" src="https://github.com/mitodl/mitxpro/assets/41048311/7e87e73f-cf62-46e1-a16d-fe230f474867">

Validation error on edit details page:
<img width="1280" alt="edit profile page 1" src="https://github.com/mitodl/mitxpro/assets/41048311/64f92bc1-4743-40a2-9c52-48f8d092bfd0">

### How can this be tested?

- Fetch, check out to this branch, install edx-api-client PR https://github.com/mitodl/edx-api-client/pull/106 with editable mode, and make sure edx is configured.
- Test the front-end validation (registration/edit details pages):
   - The full name should not start with a special character (~!@&)(+:'.?/,-), and should not contain any of (/^$#*=[]%_;<>{}|"). 
   - Test by including any of these characters, HTML, or URL in the full name.
- Test the back-end validation (registration/edit details pages):
  - Disable the front-end validation temporarily by commenting out [this line](https://github.com/mitodl/mitxpro/blame/mudassir/fix-html-urls--full-name-field/static/js/components/forms/ProfileFormFields.js#L39).
  - Check if the validation error appears by entering any URL or HTML.

- Test rate limiting/non-field errors (registration/edit details pages):
   - Disable the front-end validation as mentioned above. 
   - Set REGISTRATION_VALIDATION_RATELIMIT to a low number of requests per IP address (e.g., 10/7d). By default, the validation registration endpoint is rate-limited to 30 requests per IP address per week.
   - Ensure the notification error appears when the rate limit is exceeded.

Note: The last two pointers are for testing validation/API with backend.


### Additional Context

Further screenshots of the back-end validation:

<img width="1280" alt="notification error" src="https://github.com/mitodl/mitxpro/assets/41048311/dc419205-9a27-4167-a080-e0a184b75647">
